Add fallback resolvers for CurrentInstance

gilberto-torrezan opened the pull request:
This allow applications to inject custom default instances when the current instances cannot be found by regular means.

For example, when VaadinServlet.getCurrent() would return null, a fallback resolver could be invoked to properly create the servlet and return it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/10974)
<!-- Reviewable:end -->

gilberto-torrezan committed:
Add fallback resolvers for CurrentInstance

This allow applications to inject custom default instances when the
current instances cannot be found by regular means.

For example, when VaadinServlet.getCurrent() would return null, a
fallback resolver could be invoked to properly create the servlet and
return it.

elmot commented:
Review status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @gilberto-torrezan)

---

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10974#-LEhrBbr24Ok9oveBKSe:-LEhrBbr24Ok9oveBKSf:bw4hx2v) (no related file):*
The change needs to be tested. And the test should be clear enough to understand what is happening and why.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 64 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEhqv2E2tz6A0C89tb4:-LEhqv2E2tz6A0C89tb5:b1nlrys) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L64)):*
> ```Java
>     private static final CurrentInstance CURRENT_INSTANCE_NULL = new CurrentInstance(
>             NULL_OBJECT, true);
>     private static final Map<Class<?>, CurrentInstanceFallbackResolver<?>> fallbackResolvers = new ConcurrentHashMap<Class<?>, CurrentInstanceFallbackResolver<?>>();
> ```

The change looks pretty dangerous, there might be troubles with memory leaks, cross-session data sharing, multi-threading etc if used non-proper way. 
Which classes are going to be resolved?
Personally I do not like to expose those things into public static methods. Are there any other options to implement that functionality some other way?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

elmot commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed, 2 unresolved discussions (waiting on @gilberto-torrezan)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: all files reviewed, 4 unresolved discussions (waiting on @gilberto-torrezan)

---

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10974#-LEi3sW_9fEm2XH5YLop:-LEi3sW_9fEm2XH5YLoq:bcb89y2) (no related file):*
Shouldn't this target `master` and then separately be backported to relevant maintenance branches?

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEi46_I58uDBX1UHCBw:bdu6yub) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
> ```Java
>     public static <T> void addFallbackResolver(Class<T> type,
>             CurrentInstanceFallbackResolver<T> fallbackResolver) {
>         fallbackResolvers.put(type, fallbackResolver);
> ```

Would it make sense to guard against accidentally removing an existing fallback resolver?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: all files reviewed, 4 unresolved discussions (waiting on @gilberto-torrezan)

---

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10974#-LEi3sW_9fEm2XH5YLop:-LEiNfvZ289P9voVgh3o:brym9qm) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Shouldn't this target `master` and then separately be backported to relevant maintenance branches?
</blockquote></details>

If we start from 7, let's start from 7. If something goes wrong then it's easier to roll back the only one branch

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Make the setting of CurrentInstanceFallbackResolvers protected

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10974#-LEhrBbr24Ok9oveBKSe:-LEi_2wQE4iUER67FrmA:b7exajy) (no related file):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

The change needs to be tested. And the test should be clear enough to understand what is happening and why.
</blockquote></details>

Added an unit test.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 64 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEhqv2E2tz6A0C89tb4:-LEiZxsq0RXSNIKomSDH:b-9akgew) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L64)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

The change looks pretty dangerous, there might be troubles with memory leaks, cross-session data sharing, multi-threading etc if used non-proper way. 
Which classes are going to be resolved?
Personally I do not like to expose those things into public static methods. Are there any other options to implement that functionality some other way?

</blockquote></details>

Made the API protected, and removed the `remove` method.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEiQZZADqZAgtLuxW-N:bh0rizn) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Would it make sense to guard against accidentally removing an existing fallback resolver?
</blockquote></details>

Changed the method name to be `set` instead of `add`, to make it clear that whatever is set would remove the previously set resolver.

I don't think we need to support a chain of resolvers for each type.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEijrWx2azQjwDwDdXW:b57gt8v) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Changed the method name to be `set` instead of `add`, to make it clear that whatever is set would remove the previously set resolver.

I don't think we need to support a chain of resolvers for each type.
</blockquote></details>

Another option might to throw if there's already another handler.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 90 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikksEEZfEMKquP5hS:-LEikksEEZfEMKquP5hT:b-h9umil) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L90)):*
> ```Java
>     };
> 
>     protected CurrentInstance() {
> ```

What's the purpose of this constructor? When is it used?

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 174 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikSK610FS7PDEv5-e:-LEikSK610FS7PDEv5-f:br8dzj5) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L174)):*
> ```Java
>      * @since
>      */
>     protected static <T> void setFallbackResolver(Class<T> type,
> ```

How can anyone use this API when it's `protected`? Or seen from the other perspective: what is gained by requiring that anyone calling this API uses `com.vaadin.util` as their package name?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Remove the default constructor. Improve test.

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @elmot and @gilberto-torrezan)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEim_Hf4dVfXDhoU2bL:b-fgmvpo) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Another option might to throw if there's already another handler.
</blockquote></details>

That would make it too bureaucratic to replace a resolver (would need to remove it first).

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 90 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikksEEZfEMKquP5hS:-LEimGwJ8G0PMoXwOWyl:b-3mg8ht) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L90)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What's the purpose of this constructor? When is it used?
</blockquote></details>


Removed it and changed the current constructor to be `protected`.
This is used to allow subclassing, and calling the `setFallbackResolver` without reflection hacks.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 174 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikSK610FS7PDEv5-e:-LEilXw61pI6BYFmEo-g:b-kc4gxa) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L174)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

How can anyone use this API when it's `protected`? Or seen from the other perspective: what is gained by requiring that anyone calling this API uses `com.vaadin.util` as their package name?
</blockquote></details>

This was changed from `public` to `protected` to address the concerns pointed out by @elmot . This API is not supposed to be open to the regular users.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @elmot and @gilberto-torrezan)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEmhAlz2wY2FX32pzu2:b4qn8i7) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What's the use case where you explicitly want to replace a resolver?
</blockquote></details>

No use case at all, since this feature is a very specific one. I just don't want to make it too cumbersome for no reason. Making a setter throw if there's already a value over there feels very unintuitive IMO.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 90 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikksEEZfEMKquP5hS:-LEmiAzO9kfMDT5PT1g6:b-7ngjjh) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L90)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What's the purpose of subclassing `CurrentInstance`?
</blockquote></details>

To be able to call `setFallbackResolver` (which is protected in this revision) without reflection hacks.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 174 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikSK610FS7PDEv5-e:-LEmhUiW1v_6g11gg1m2:bbh2lxy) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L174)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Nothing in the `CurrentInstance` class is not supposed to be used by regular users, so I don't see why this particular method would be any different from all the other methods?
</blockquote></details>

I have no strong opinions about it. Now addressing what @elmot mentioned in his comment, what else can be done to avoid misuse of the API if it is made public?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Made setFallbackResolver public again

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 64 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEhqv2E2tz6A0C89tb4:-LEmuA-n4T9Ju2VmmD6T:b-snzgyc) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L64)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Made the API protected, and removed the `remove` method.
</blockquote></details>

Rolled back to `public`.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 90 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikksEEZfEMKquP5hS:-LEmtvKg-xt5yiffzAq9:b-7jktbc) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L90)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

To be able to call `setFallbackResolver` (which is protected in this revision) without reflection hacks.
</blockquote></details>

`setFallbackResolver` was set as public again, now the constructor was set back to `private`.

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 174 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikSK610FS7PDEv5-e:-LEmu1a_-ykF2xESHwGo:b-e1xu8r) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L174)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

I have no strong opinions about it. Now addressing what @elmot mentioned in his comment, what else can be done to avoid misuse of the API if it is made public?
</blockquote></details>

Talked with @elmot offline, he agreed to make the method `public`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEmsolm11QpYra4VuJZ:b-n935md) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
What's the problem with making non-existent use cases cumbersome? Handling this case matters only if the same application would use two add-ons that try to register a resolver for the same type. 

Silently discarding either resolver would most likely cause one of the add-ons to misbehave in a way that would be really hard for the application developer to debug. To make things wore, the exact outcome would be different depending on which add-on happens to be initialized first. In what situation would that be desirable?

Throwing makes the application developer aware of the conflict.

> Making a setter throw if there's already a value over there feels very unintuitive IMO.

Then rename it to `defineFallbackResolver`?

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 174 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikSK610FS7PDEv5-e:-LEmrHZo3Bxi_XUMxUsu:b-z755sx) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L174)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

I have no strong opinions about it. Now addressing what @elmot mentioned in his comment, what else can be done to avoid misuse of the API if it is made public?
</blockquote></details>

If someone wants to misuse it, they will be able to do that no matter what we do by using e.g. reflection, putting files in the same package or messing with subclasses.

The risk of accidental misuse is already quite minimal since `CurrentInstance` is only an internal matter that application developers aren't supposed to directly interact with. There are instead specific accessors such as `UI.getCurrent()` that work as the application developer API.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Rename the method to defineFallbackResolver, and make it throw when a
type is used twice

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot and @gilberto-torrezan)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 171 at r1](https://reviewable.io/reviews/vaadin/framework/10974#-LEi46_I58uDBX1UHCBv:-LEnAo4qA4ryGfM_Z59U:b-i7nqd) ([raw file](https://github.com/vaadin/framework/blob/5f18e34e5670e17aba1d452069a7bb66996d196b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L171)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What's the problem with making non-existent use cases cumbersome? Handling this case matters only if the same application would use two add-ons that try to register a resolver for the same type. 

Silently discarding either resolver would most likely cause one of the add-ons to misbehave in a way that would be really hard for the application developer to debug. To make things wore, the exact outcome would be different depending on which add-on happens to be initialized first. In what situation would that be desirable?

Throwing makes the application developer aware of the conflict.

> Making a setter throw if there's already a value over there feels very unintuitive IMO.

Then rename it to `defineFallbackResolver`?
</blockquote></details>

Renamed and made it throw. Added unit tests for it as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot and @gilberto-torrezan)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 175 at r5](https://reviewable.io/reviews/vaadin/framework/10974#-LEnLLVP07jJwe-Fd-GV:-LEnLLVQ5p3xjsW66bkJ:b-3iirnr) ([raw file](https://github.com/vaadin/framework/blob/36da4eec2ebcdccb7f119fbfcb07645ef5dbd47b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L175)):*
> ```Java
>     public static <T> void defineFallbackResolver(Class<T> type,
>             CurrentInstanceFallbackResolver<T> fallbackResolver) {
>         if (fallbackResolvers.containsKey(type)) {
> ```

This is not threadsafe. Another thread might have put a resolver there between when the original thread runs `containsKey` and `put`. Should instead use `putIfAbsent` and check the return value.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Make the method thread-safe

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 175 at r5](https://reviewable.io/reviews/vaadin/framework/10974#-LEnLLVP07jJwe-Fd-GV:-LEnYJbI8XIljTU3ZJ9w:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/36da4eec2ebcdccb7f119fbfcb07645ef5dbd47b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L175)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

This is not threadsafe. Another thread might have put a resolver there between when the original thread runs `containsKey` and `put`. Should instead use `putIfAbsent` and check the return value.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 175 at r5](https://reviewable.io/reviews/vaadin/framework/10974#-LEnLLVP07jJwe-Fd-GV:-LEnZgpk9gU-OzqgSs3Y:b-i1ybn5) ([raw file](https://github.com/vaadin/framework/blob/36da4eec2ebcdccb7f119fbfcb07645ef5dbd47b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L175)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Actually not done. `putIfAbsent` is a Java 8 feature.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Make the method thread-safe in a Java 6 way

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @elmot, @gilberto-torrezan, and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 175 at r5](https://reviewable.io/reviews/vaadin/framework/10974#-LEnLLVP07jJwe-Fd-GV:-LEnadO00jPUo8CGbW9c:b-p6cpez) ([raw file](https://github.com/vaadin/framework/blob/36da4eec2ebcdccb7f119fbfcb07645ef5dbd47b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L175)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Actually not done. `putIfAbsent` is a Java 8 feature.
</blockquote></details>

Now done in an old-fashioned way.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @elmot and @Legioth)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 175 at r5](https://reviewable.io/reviews/vaadin/framework/10974#-LEnLLVP07jJwe-Fd-GV:-LEneeUp-JHB3zuTV-i4:b58u3r4) ([raw file](https://github.com/vaadin/framework/blob/36da4eec2ebcdccb7f119fbfcb07645ef5dbd47b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L175)):*
> `putIfAbsent` is a Java 8 feature.

The javadocs for Java 6 are of a differing opinion. https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/ConcurrentHashMap.html#putIfAbsent(K,%20V)

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Thread safety with ConcurrentHashMap API instead of just Map

gilberto-torrezan commented:
Review status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @elmot)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 175 at r5](https://reviewable.io/reviews/vaadin/framework/10974#-LEnLLVP07jJwe-Fd-GV:-LEnhR1N2LouOyz_yoAB:blomt4b) ([raw file](https://github.com/vaadin/framework/blob/36da4eec2ebcdccb7f119fbfcb07645ef5dbd47b/server/src/main/java/com/vaadin/util/CurrentInstance.java#L175)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

> `putIfAbsent` is a Java 8 feature.

The javadocs for Java 6 are of a differing opinion. https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/ConcurrentHashMap.html#putIfAbsent(K,%20V)
</blockquote></details>

Ok, so that's a Java 8 feature on the Map interface then.
Rolling back...

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

elmot commented:
Reviewed 2 of 3 files at r2, 1 of 1 files at r8.
Review status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @elmot and @gilberto-torrezan)

---

*[server/src/main/java/com/vaadin/util/CurrentInstance.java, line 90 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEikksEEZfEMKquP5hS:-LEms5ae572k-QtecIaA:b7blp48) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/main/java/com/vaadin/util/CurrentInstance.java#L90)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

To be able to call `setFallbackResolver` (which is protected in this revision) without reflection hacks.
</blockquote></details>

We are trying to hide a little bit dangerous API (i.e. make it protected and callable only from subclasses). This is the price of it.

---

*[server/src/test/java/com/vaadin/util/CurrentInstanceTest.java, line 246 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEnkFC84CmhpfrPFLkP:-LEnkFC84CmhpfrPFLkQ:bbkm1zm) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/test/java/com/vaadin/util/CurrentInstanceTest.java#L246)):*
> ```Java
>                 serviceResolver);
> 
>         VaadinService.getCurrent();
> ```

The result is not verified.

---

*[server/src/test/java/com/vaadin/util/CurrentInstanceTest.java, line 279 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEnkOSB-ScqBckRAETh:-LEnkOSB-ScqBckRAETi:b-4ds3m8) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/test/java/com/vaadin/util/CurrentInstanceTest.java#L279)):*
> ```Java
>         public T resolve() {
>             called++;
>             return null;
> ```

Returning null is a bit strange for the test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

elmot commented:
Reviewed 2 of 3 files at r5.
Review status: all files reviewed, 2 unresolved discussions (waiting on @gilberto-torrezan)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: all files reviewed, 2 unresolved discussions (waiting on @gilberto-torrezan)

---

*[server/src/test/java/com/vaadin/util/CurrentInstanceTest.java, line 279 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEnkOSB-ScqBckRAETh:-LEnl6ra8HxPrXxqWDW6:b1x1yn5) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/test/java/com/vaadin/util/CurrentInstanceTest.java#L279)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Returning null is a bit strange for the test.
</blockquote></details>

What is returned doesn't really matter, as long as the resolver is called.
I wanted to avoid creating Mockup fake classes for every one of the cases (UI, Servlet, Service...)

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Improve test with fake classes.

gilberto-torrezan commented:
Review status: all files reviewed, 2 unresolved discussions (waiting on @elmot)

---

*[server/src/test/java/com/vaadin/util/CurrentInstanceTest.java, line 246 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEnkFC84CmhpfrPFLkP:-LEnq2ZC8zOMNk31_J1w:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/test/java/com/vaadin/util/CurrentInstanceTest.java#L246)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

The result is not verified.
</blockquote></details>

Done.

---

*[server/src/test/java/com/vaadin/util/CurrentInstanceTest.java, line 279 at r2](https://reviewable.io/reviews/vaadin/framework/10974#-LEnkOSB-ScqBckRAETh:-LEnq246CR5K1fsNNNlg:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/674b62f35afa60b8fed7f14602b42d1623492c86/server/src/test/java/com/vaadin/util/CurrentInstanceTest.java#L279)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

What is returned doesn't really matter, as long as the resolver is called.
I wanted to avoid creating Mockup fake classes for every one of the cases (UI, Servlet, Service...)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

elmot commented:
Reviewed 1 of 1 files at r9.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan committed:
Clear the test state after it has been run.

tsuoanttila commented:
Reviewed 1 of 1 files at r10.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/framework/10974)*
<!-- Sent from Reviewable.io -->
